File: Dynamic_programing/494.target-sum.ts
// @leet start
function findTargetSumWays(nums: number[], target: number): number {
  const memo = new Map<string, number>();
  function dfs(level: number, cur: number): number {
    const key = `${level}:${cur}`;
    if (memo.has(key)) return memo.get(key)!;
    if (level == nums.length) {
      return target == cur ? 1 : 0;
    }
    const l = dfs(level + 1, cur + nums[level]);
    const r = dfs(level + 1, cur - nums[level]);
    memo.set(key, l + r);
    return l + r;
  }
  return dfs(0, 0);
}
// @leet end
function findTargetSumWays2(nums: number[], target: number): number {
  let sum = 0;
  for (const i of nums) {
    sum += i;
  }
  if (Math.abs(target) > sum) return 0;
  if ((target + sum) % 2 == 1) return 0;
  const total = Math.floor((target + sum) / 2);
  const dp = Array(total + 1).fill(0);
  dp[0] = 1;
  for (const item of nums) {
    for (let j = dp.length - 1; j >= item; j--) {
      dp[j] += dp[j - item];
    }
  }
  return dp[dp.length - 1];
}

File: Dynamic_programing/494.target-sum.py
from typing import List
from functools import cache


# @leet start
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        @cache
        def dfs(level, cur):
            if level == len(nums):
                return 1 if cur == target else 0
            l = dfs(level + 1, cur + nums[level])
            r = dfs(level + 1, cur - nums[level])
            return l + r

        return dfs(0, 0)


# @leet end
class Solution2:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums) + target
        if abs(target) > sum(nums):
            return 0
        if total % 2 == 1:
            return 0
        total = total // 2
        dp = [1] + [0] * total
        for item in nums:
            for j in range(len(dp) - 1, item - 1, -1):
                # if dp[j - item] > 0: ## no nessary, but helpful to keep for df
                dp[j] += dp[j - item]
        return dp[-1]

